app/routes.py

```
@@ -1 +1,371 @@
+
+    goal = Goal(title=title)
+    db.session.add(goal) 
+    db.session.commit()  
+
+    response = {
+        "goal": {
+            "id": goal.goal_id,  
+            "title": goal.title
+        }
+    }
+
+    return make_response(jsonify(response), 201)
+
+
+@goals_bp.route("/<goal_id>", methods=["PUT"])
+def update_goal(goal_id):
+    goal = validate_model(Goal, goal_id)
+    request_body = request.get_json()
+
+    if not request_body:
+        return make_response({"details": "Invalid data"}, 400)
+
+    title = request_body.get("title")
+    description = request_body.get("description")
+
+    if not title or not description:
+        return make_response({"details": "Invalid data"}, 400)
+
+    goal.title = title
+    goal.description = description
+
+    db.session.commit()
+
+    response = {
+        "goal": {
+            "goal_id": goal.goal_id,
+            "title": goal.title,
+            "description": goal.description,
+            "is_complete": False
+        }
+    }
+
+    return make_response(jsonify(response), 200)
+
+
+
+
+@goals_bp.route("/<goal_id>", methods=["DELETE"])
+def delete_goal(goal_id):
+    goal = validate_model(Goal, goal_id)
+
+    db.session.delete(goal)
+    db.session.commit()
+
+    return make_response({"details": f"Goal {goal_id} \"{goal.title}\" successfully deleted"}, 200)
+
+#relationships start here:
+# 
+@goals_bp.route("/<goal_id>/tasks", methods=["POST"])
+
+def associate_tasks_with_goal(goal_id):
+    goal = validate_model(Goal, goal_id)
+    request_body = request.get_json()
+
+    if not request_body:
+        return make_response({"details": "Invalid data"}, 400)
+
+    task_ids = request_body.get("task_ids")
+
+    if not task_ids or not isinstance(task_ids, list):
+        return make_response({"details": "Invalid data"}, 400)
+
+    tasks = Task.query.filter(Task.task_id.in_(task_ids)).all()
+
+    if len(tasks) != len(task_ids):
+        return make_response({"details": "Invalid task IDs"}, 400)
```

👍 Nice way to retrieve a group of tasks based on their ids, but notice this has the effect of silently ignoring invalid task ids. Another way we might approach this is
```py
    tasks = [validate_model(Task, task_id) for task id in task_ids]
```
which would abort the request if we received a bad task id.